### PR TITLE
[Error] Raise an warning when a Taichi matrix have > 32 entries

### PR DESCRIPTION
**Pull request by archibate**
Related issue = WeChat group

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
```py
import taichi as ti
import time

for N in range(1, 20):
    M = 60 // N

    ti.init()
    ti.core.toggle_advanced_optimization(False)

    m = ti.Matrix.var(N, N, ti.f32, M)
    n = ti.Matrix.var(N, N, ti.f32, M)
    a = ti.Matrix.var(N, N, ti.f32, M)

    @ti.kernel
    def func():
        for i in m:
            a[i] = m[i] + n[i]

    begin = time.time()
    func()
    end = time.time()
    print(N, end - begin)

'''
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-y5ni86_b
[Taichi] <dev mode>, llvm 10.0.0, commit 0ac65b1e, python 3.8.3
[Taichi] Starting on arch=x64
1 0.5222694873809814
[Taichi] Starting on arch=x64
2 0.5329015254974365
[Taichi] Starting on arch=x64
3 0.5772120952606201
[Taichi] Starting on arch=x64
4 0.6571242809295654
[Taichi] Starting on arch=x64
5 0.7596781253814697
[Taichi] Starting on arch=x64
6 0.9348974227905273
[Taichi] Starting on arch=x64
7 1.1725513935089111
[Taichi] Starting on arch=x64
8 1.5137403011322021
[Taichi] Starting on arch=x64
9 1.9572887420654297
[Taichi] Starting on arch=x64
10 2.5833590030670166
[Taichi] Starting on arch=x64
11 3.3029255867004395
[Taichi] Starting on arch=x64
12 4.244035720825195
[Taichi] Starting on arch=x64
13 5.5972113609313965
[Taichi] Starting on arch=x64
14 7.187188625335693
[Taichi] Starting on arch=x64
15 9.507669448852539
[Taichi] Starting on arch=x64
16 12.00432562828064
[Taichi] Starting on arch=x64
17 15.767349243164062
[Taichi] Starting on arch=x64
18 20.062804222106934
[Taichi] Starting on arch=x64
[E 07/04/20 22:24:04.456] [struct_llvm.cpp:run@220] (int)snodes.size() <= taichi_max_num_snodes


***********************************
* Taichi Compiler Stack Traceback *
***********************************
/tmp/taichi-y5ni86_b/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
/tmp/taichi-y5ni86_b/taichi_core.so: taichi::lang::StructCompilerLLVM::run(taichi::lang::SNode&, bool)
/tmp/taichi-y5ni86_b/taichi_core.so: taichi::lang::Program::materialize_layout()
/tmp/taichi-y5ni86_b/taichi_core.so: taichi::lang::layout(std::function<void ()> const&)
/tmp/taichi-y5ni86_b/taichi_core.so(+0x8e300f) [0x7f995155c00f]
/tmp/taichi-y5ni86_b/taichi_core.so(+0x7ba228) [0x7f9951433228]
/usr/lib/libpython3.8.so.1.0: PyCFunction_Call
/usr/lib/libpython3.8.so.1.0: _PyObject_MakeTpCall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0(+0x13e442) [0x7f9961bf4442]
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyObject_FastCallDict
/usr/lib/libpython3.8.so.1.0: _PyObject_Call_Prepend
/usr/lib/libpython3.8.so.1.0(+0x1f5e09) [0x7f9961cabe09]
/usr/lib/libpython3.8.so.1.0: PyObject_Call
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: _PyFunction_Vectorcall
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalFrameDefault
/usr/lib/libpython3.8.so.1.0: _PyEval_EvalCodeWithName
/usr/lib/libpython3.8.so.1.0: PyEval_EvalCode
/usr/lib/libpython3.8.so.1.0(+0x1d8248) [0x7f9961c8e248]
/usr/lib/libpython3.8.so.1.0(+0x1d2483) [0x7f9961c88483]
/usr/lib/libpython3.8.so.1.0: PyRun_FileExFlags
/usr/lib/libpython3.8.so.1.0: PyRun_SimpleFileExFlags
/usr/lib/libpython3.8.so.1.0: Py_RunMain
/usr/lib/libpython3.8.so.1.0: Py_BytesMain
/usr/lib/libc.so.6: __libc_start_main
python(_start+0x2e) [0x55646879504e]

Internal Error occurred, check this page for possible solutions:
https://taichi.readthedocs.io/en/stable/install.html#troubleshooting
Traceback (most recent call last):
  File "w.py", line 20, in <module>
    func()
  File "/root/taichi/python/taichi/lang/kernel.py", line 553, in wrapped
    return primal(*args, **kwargs)
  File "/root/taichi/python/taichi/lang/kernel.py", line 483, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/root/taichi/python/taichi/lang/kernel.py", line 285, in materialize
    self.runtime.materialize()
  File "/root/taichi/python/taichi/lang/impl.py", line 182, in materialize
    taichi_lang_core.layout(layout)
RuntimeError: [struct_llvm.cpp:run@220] (int)snodes.size() <= taichi_max_num_snodes
'''
```

### CODE REVIEW

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -89,42 +92,52 @@ def __init__(self,
...
+                            var_list.append(e.grad)
+                    ti.root.dense(ti.index_nd(dim),
+                                  shape).place(*tuple(var_list), offset=offset)
+            else:
+                assert offset is None, f"shape cannot be None when offset is being set"
+
+        if self.n * self.m > 100:
+            raise TaichiSyntaxError(
+                f'Taichi matrices/vectors with {self.n}x{self.m} > 100 entries are not suggested.'
+                ' Matrices/vectors will be automatically unrolled at compile-time for performance.'
```
Comment (Rullec) [inline python/taichi/lang/matrix.py:135]:
```suggestion
                ' Matrices/Vectors will be automatically unrolled at compile-time for performance.'
```

**Comment by Rullec** [comment]
Nice work, LGTM and a nit! :)
Btw, why we use `TaichiSyntaxError` here? The attempt for trying to allocate a big matrix is not an illegal syntax I guess.

IMO I would suggest `ValueError` coming from [python doc](https://docs.python.org/3/library/exceptions.html#ValueError)
> exception ValueError
> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.

**Comment by yuanming-hu** [comment]
Thanks for implementing this. I don't think it should be an error since it's not a fundamental limitation of the system. We should probably use a warning instead.

**Comment by archibate** [comment]
> Thanks for implementing this. I don't think it should be an error since it's not a fundamental limitation of the system. We should probably use a warning instead.

Done. It's an `UserWarning` now.

File: misc/prtags.json [hunk 2]
```diff
@@ -29,5 +29,6 @@
   "perf"            : "Performance improvements",
   "ipython"         : "IPython and other shells",
   "cc"              : "C source backend",
+  "error"           : "Error message",
```
Comment (yuanming-hu) [inline misc/prtags.json:32]:
```suggestion
  "error"           : "Error messages",
```

File: python/taichi/lang/matrix.py [hunk 3]
```diff
@@ -90,42 +92,54 @@ def __init__(self,
...
+                    offset = (offset, )
 
-            if offset is not None:
-                assert len(shape) == len(
-                    offset
-                ), f'The dimensionality of shape and offset must be the same  (f{len(shape)} != f{len(offset)})'
+                if offset is not None:
+                    assert len(shape) == len(
+                        offset
+                    ), f'The dimensionality of shape and offset must be the same  (f{len(shape)} != f{len(offset)})'
```
Comment (archibate) [inline python/taichi/lang/matrix.py:106]:
```suggestion
                    ), f'The dimensionality of shape and offset must be the same  ({len(shape)} != {len(offset)})'
```
@rexwangcc ?

**Pull request merged**